### PR TITLE
Override deploy:shared task to handle pimcore php config files

### PR DESCRIPTION
**Pull request by dlhck**

### CODE REVIEW

File: recipes/pimcore.php [hunk 1]
```diff
@@ -12,10 +12,85 @@
...
 task('deploy:pimcore:install-classes', function() {
     run('{{bin/php}} {{bin/console}} deployment:classes-rebuild -c');
 });
 
 task('deploy:pimcore:migrate', function() {
     run('{{bin/php}} {{bin/console}} pimcore:migrations:migrate');
+});
+
+desc('Creating symlinks for shared files and dirs');
+task('deploy:shared', function () {
```
Comment (dpfaffenbauer) [inline recipes/pimcore.php]: can we rename that to ```deploy:pimcore:shared:config``` and introduce another configuration ```pimcore:shared_configurations``` to configure them? I don't think it as a good idea to have them static and overwrite a default function of deployer.

File: recipes/pimcore.php [hunk 2]
```diff
@@ -12,10 +12,85 @@
...
+    $sharedPath = "{{deploy_path}}/shared";
+    // Validate shared_dir, find duplicates
+    foreach (get('shared_dirs') as $a) {
+        foreach (get('shared_dirs') as $b) {
+            if ($a !== $b && strpos(rtrim($a, '/') . '/', rtrim($b, '/') . '/') === 0) {
+                throw new Exception("Can not share same dirs `$a` and `$b`.");
+            }
+        }
+    }
+    foreach (get('shared_dirs') as $dir) {
```
Comment (dpfaffenbauer) [inline recipes/pimcore.php]: remove this completely as we introduce a new task

File: recipes/pimcore.php [hunk 3]
```diff
@@ -12,10 +12,85 @@
...
+        }
+        // Remove from source.
+        run("rm -rf {{release_path}}/$dir");
+        // Create path to shared dir in release dir if it does not exist.
+        // Symlink will not create the path and will fail otherwise.
+        run("mkdir -p `dirname {{release_path}}/$dir`");
+        // Symlink shared dir to release dir
+        run("{{bin/symlink}} $sharedPath/$dir {{release_path}}/$dir");
+    }
+    foreach (get('shared_files') as $file) {
```
Comment (dpfaffenbauer) [inline recipes/pimcore.php]: use ```pimcore:shared_configurations```

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]

File: deploy.sample.php [hunk 4]
```diff
@@ -45,6 +43,16 @@
     'var/versions',
     'var/sessions'
 ]);
+// the configuration files of pimcore that will be processed at creation
+set('pimcore_shared_configurations', [
+    'var/config/website-settings.php',
+    'var/config/reports.php',
+    'var/config/extensions.php',
```
Comment (dpfaffenbauer) [inline deploy.sample.php]: can you please remove the extensions.php? this one should come from the repository.

File: recipes/pimcore.php [hunk 5]
```diff
@@ -12,10 +12,37 @@
...
+
+// process the pimcore config files
+// add empty array if file is empty
+task('deploy:pimcore:shared:config', function () {
+
+    $sharedPath = "{{deploy_path}}/shared";
+    $emptyContent = "<?php return [];";
+
+    foreach (get('pimcore_shared_configurations') as $configFile) {
+        run("if [ grep -q '[^[:space:]]' < '$sharedPath/$configFile' ]; then echo '$emptyContent' >> '$sharedPath/$configFile'; fi");
```
Comment (dpfaffenbauer) [inline recipes/pimcore.php]: this doesn't seem to work for me, but this did:

```php
run("[ -s '$sharedPath/$configFile' ] || echo '$emptyContent' >> '$sharedPath/$configFile'");
```

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]

File: deploy.sample.php [hunk 6]
```diff
@@ -45,6 +43,16 @@
...
+    'var/config/web2print.php',
+    'var/config/workflowmanagement.php',
+    'var/config/perspectives.php',
+    'var/config/customviews.php'
```
Comment (dpfaffenbauer) [inline deploy.sample.php]: this shouldn't be there as well, causes an pimcore failure when it returns empty array

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]

File: deploy.sample.php (hunk 6 above)
Comment (dpfaffenbauer) [reply deploy.sample.php]: same for perspectives.... doesn't show any tree views (documents, assets, objects) when empty

**Review by dpfaffenbauer — APPROVED** [review]

**Comment by aarongerig** [comment]
@davidhoeck Thanks for the effort and @dpfaffenbauer thanks for reviewing! ;)

**Pull request merged**